clients/src/main/java/org/apache/kafka/common/record/Record.java

```diff
@@ -434,4 +311,268 @@ public int hashCode() {
         return buffer.hashCode();
     }
 
+    public int convertedSize(byte toMagic) {
+        return recordSize(toMagic, Math.max(0, keySize()), Math.max(0, valueSize()));
+    }
+
+    public Record convert(byte toMagic) {
+        if (toMagic == magic())
+            return this;
+
+        ByteBuffer buffer = ByteBuffer.allocate(convertedSize(toMagic));
+        TimestampType timestampType = wrapperRecordTimestampType != null ?
+                wrapperRecordTimestampType : TimestampType.forAttributes(attributes());
+        convertTo(buffer, toMagic, NO_TIMESTAMP, timestampType);
```

Comment:
To be future proof, should we pass in the timestamp in this record instead of NO_TIMESTAMP?